Refactor unittests for activation functions relu, elu, and sigmoid

paulshaoyuqiao opened the pull request:
Summary: The tests covered previously by test_qrelu, test_qelu, and test_qsigmoid are now merged into one test to ensure conciseness and reduce redundancy.

Test Plan:
1. On a devserver, build PyTorch from source by running the command "buck build mode/dev //caffe2:torch"
 2. Run the merged unit test throught the command "buck test mode/dev //caffe2/test:quantization -- test_qrelu_elu_sigmoid

Differential Revision: D21755690

vkuzo commented:
would it be possible to share the code but still have separate tests per activation function?  That way, if one of the activation tests is flaky / disabled / etc, it wouldn't impact the other tests

vkuzo commented:
> would it be possible to share the code but still have separate tests per activation function? That way, if one of the activation tests is flaky / disabled / etc, it wouldn't impact the other tests

it could be something like 

```
activation_test_config = { ... }

def _test_activation(...):
  ...the shared test logic...

# one test case for each activation
test_foo(self, ...):
    foo_config = activation_test_config['foo']
    _test_activation(foo_config)

```

z-a-f commented on test/quantization/test_quantized_op.py:
```diff
@@ -118,68 +118,172 @@ def _get_random_tensor_and_q_params(shapes, rand_scale, torch_type):
+            }
+        ],
+        'relu6': [
+            {
+                'quantized_fn': [
+                    torch.ops.quantized.relu6
+                ],
+                'reference_fn': torch.nn.functional.relu6
+            },
+            {
+                'quantized_fn': [
+                    torch.nn.quantized.ReLU6()
+                ],
+                'reference_fn': torch.nn.ReLU6(),
+            },
+            {
+                'quantized_fn': [
+                    torch.nn.quantized.ReLU6(inplace=True)
+                ],
+                'reference_fn': torch.nn.ReLU6(),
+            }
```
nit: It looks to me this configurations are very similar. Can you collapse them into one, and just add all of them to the list of `quantized_fn`?

```python
           {
                'quantized_fn': [
                    torch.ops.quantized.relu6,
                    torch.nn.quantized.ReLU6(inplace=False),
                    torch.nn.quantized.ReLU6(inplace=True)
                ],
                'reference_fn': torch.nn.functional.relu6
            },
```

z-a-f commented on test/quantization/test_quantized_op.py:
```diff
@@ -118,68 +118,172 @@ def _get_random_tensor_and_q_params(shapes, rand_scale, torch_type):
+    # no specified, the range will be (-inf, inf)
+    activation_test_config = {
+        'relu': [
+            {
+                'quantized_fn': [
+                    torch.relu
+                ],
+                'reference_fn': torch.relu
+            },
+            {
+                'quantized_fn': [
+                    torch.relu_
+                ],
+                'reference_fn': torch.relu_,
+            },
```
nit: Can you also merge these two configurations (see example below).

z-a-f commented on test/quantization/test_quantized_op.py:
```diff
@@ -118,68 +118,172 @@ def _get_random_tensor_and_q_params(shapes, rand_scale, torch_type):
+    # quantized function is assumed to be either inplace by default or the
+    # test is not testing an inplace function.
+    # output_range: the output range the operator will map to. By default, if it is
+    # no specified, the range will be (-inf, inf)
```
nit: Is the range [-inf, inf] or is it dependent on the X min max?

z-a-f reviewed (approved):
LGTM, less some nits

jerryzh168 commented on test/quantization/test_quantized_op.py:
```diff
@@ -118,68 +118,172 @@ def _get_random_tensor_and_q_params(shapes, rand_scale, torch_type):
+                    # Adjusts output_zero_point if needed (for sigmoid and hardsigmoid).
+                    # output_zero_point determines the additional offset that will be
+                    # added to a scaled value during quantization.
+                    if fn_name in ['sigmoid', 'hardsigmoid']:
```
can you make this a config as well?

jerryzh168 commented on test/quantization/test_quantized_op.py:
```diff
@@ -118,68 +118,172 @@ def _get_random_tensor_and_q_params(shapes, rand_scale, torch_type):
+    # test is not testing an inplace function.
+    # output_range: the output range the operator will map to. By default, if it is
+    # no specified, the range will be (-inf, inf)
+    activation_test_config = {
```
I feel moving the config to the tests would be better?

z-a-f replied:
Do you mean make it part of the helper function below (`_test_activation_function`)? If so, we were originally planning on doing it, but I think it just bloats thee method. Also, we were thinking in the future we can separate the activations functions into separate class, say `TestQuantizedActivationsOps`, which would be configuration based only. In leu of the that, we kinda converged to the class-level config. Do you think we should move it inside the helper method still?

jerryzh168 replied:
I mean put this in individual tests, like `test_qrelu`, that way it is easier to see what is been tested from the test itself

jerryzh168 replied:
I think putting in individual tests would mean one less lookup, not sure what is the benefit of putting this in class level?

paulshaoyuqiao replied:
That makes sense. By distributing the config individually into each of the test function, they can be changed more easily and directly. I will add that change!

jerryzh168 commented on test/quantization/test_quantized_op.py:
```diff
@@ -118,68 +118,164 @@ def _get_random_tensor_and_q_params(shapes, rand_scale, torch_type):
-            'native': torch.relu,
-            'nn.functional': torch.nn.functional.relu,
-        }
+        op_groups = test_configs
```
nit: remove `test_configs` or rename `op_groups` to `test_config`

jerryzh168 commented on test/quantization/test_quantized_op.py:
```diff
@@ -118,68 +118,164 @@ def _get_random_tensor_and_q_params(shapes, rand_scale, torch_type):
-    @given(X=hu.tensor(shapes=hu.array_shapes(1, 5, 1, 5),
-                       qparams=hu.qparams()))
-    def test_qrelu(self, X):
+    # When writing a unit test for the activation function,
```
nit: use doc string for this comment? and move this to function `_test_activation_function`?

jerryzh168 reviewed (approved):
LGTM, had a few nit comments

Commit:
Refactor unittests for activation functions relu, elu, and sigmoid (#39190)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/39190

The tests covered previously by test_qrelu, test_qrelu6, test_qsigmoid, and test_qhardsigmoid are now merged into one test to ensure conciseness and reduce redundancy.

The refactoring aims to provide the basis for a more generalizable framework to test quantized activation functions and more in the future.

Test Plan:
1. On a devserver, build PyTorch from source by running the command "buck build mode/dev //caffe2:torch"
 2. Run the merged unit test throught the command
"buck test mode/dev //caffe2/test:quantization -- test_qrelu"
"buck test mode/dev //caffe2/test:quantization -- test_qrelu6"
"buck test mode/dev //caffe2/test:quantization -- test_qsigmoid"
"buck test mode/dev //caffe2/test:quantization -- test_qhardsigmoid"

Reviewed By: z-a-f

Differential Revision: D21755690

fbshipit-source-id: 9e8296a1dadd71c4bd40ec61555fa4e42aac818a
